cpg_infra/config/deserializabledataclass.py

```diff
@@ -30,94 +30,110 @@ def __post_init__(self):
-                    break
-                except TypeError as exc:
-                    e = exc
-
-            if e:
-                raise e
+            try:
+                self.__dict__[fieldname] = try_parse_value_as_type(value, ftype)
+            except ValueError as e:
+                raise ValueError(
+                    f'Error parsing {self.__class__.__name__}.{fieldname} :: {value!r}'
+                ) from e
 
     def __repr__(self):
         args = ', '.join(f'{k}={v!r}' for k, v in vars(self).items())
         return f'{self.__class__.__name__}( {args} )'
 
 
-def parse_value_from_type(config, fieldname, ftype):
-    if ftype is None:
-        return None
-
-    if ftype in (list, dict) or get_origin(ftype) in (list, dict):
-        ftype_type = ftype if ftype in (list, dict) else get_origin(ftype)
-        value = config.get_object(fieldname)
+def try_parse_value_as_type(value, dtype):
+    """
+    Try to parse a value as a specific type, if it fails
+    return None
+    """
+    if dtype is None or isinstance(dtype, type(None)):
+        if value is None:
+            return None
+        raise ValueError(f'Expected None, got {type(value)} for {value!r}')
+
+    if isinstance(dtype, UnionType):
+        dtype = get_args(dtype)
+
+    if isinstance(dtype, tuple):
+        # union type
+        if len(dtype) == 0:
+            # any
+            return value
 
-        if value and isinstance(value, ftype_type):
+        # union type
+        if value is None:
+            if any(isinstance(t, type(None)) for t in dtype):
+                return None
+            raise ValueError(f'Expected (non-optional) {dtype}, got None')
+
+        union_parse_errors = []
+        for t in dtype:
+            try:
+                return try_parse_value_as_type(value, t)
+            except ValueError as e:
+                # debugging
+                if not (t is None or isinstance(t, type(None))):
+                    # skip optional errors
+                    union_parse_errors.append(e)
+                continue
+        if len(union_parse_errors) == 1:
+            message = f'Could not coerce value of type ({type(value)!r}) as any of the types in the union: {dtype}, value :: {value!r}'
+            raise ValueError(message) from union_parse_errors[0]
+        error_message = ''.join([f'\n\t{e}' for e in union_parse_errors])
+        message = (
+            f'Could not coerce value of type ({type(value)!r}) as any of the types '
+            f'in the union: {dtype}, value :: {value!r}, from errors: {error_message}'
+        )
+        raise ValueError(message)
+
+    if isinstance(dtype, tuple) and len(dtype) == 1:
+        return try_parse_value_as_type(value, dtype[0])
+
+    if dtype is list or get_origin(dtype) is list:
+        if not isinstance(value, list):
+            raise ValueError(f'Expected list, got {type(value)} for {value!r}')
+        list_types = get_args(dtype)
+        if len(list_types) != 1:
```

Comment:
Is there a use-case where you would want to validate something like `list[a | b | c | ... ]` to make sure that list items only have types `a | b | c | ...` ?

Reply:
Yes! I've added a test to avoid regressions of this case :)

Reply:
Ah I see the confusion, the type here from get_args for a list is always one element, and would come in as a UnionType which is caught below.

Some other types, like dict[str, str] have get_args returning 2 elements.